File: transformation/src/dynamic/scripts/dynamic_s.py
#! /usr/bin/env python
import rospy
import tf2_ros#创建订阅对象所需包含的头文件
from tf2_geometry_msgs import tf2_geometry_msgs#组织被转换的坐标点所需包含的头文件
"""  
    订阅坐标系信息，生成一个相对于 子级坐标系的坐标点数据，
    转换成父级坐标系中的坐标点

    实现流程:
        1.导包
        2.初始化 ROS 节点
        3.创建 TF 订阅对象
        4.创建一个 radar 坐标系中的坐标点
        5.调研订阅对象的 API 将 4 中的点坐标转换成相对于 world 的坐标
        6.spin

"""


if __name__=="__main__":
    rospy.init_node("static_sp")#初始化节点
    #创建订阅对象
    buffer=tf2_ros.Buffer()#创建缓存对象
    sub=tf2_ros.TransformListener(buffer)#创建订阅对象（将所订阅的数据传入缓存中）
    #组织被转换的坐标点
    ps=tf2_geometry_msgs.PointStamped()#组织被发布消息【参看坐标msgs消息】
    ps.header.stamp=rospy.Time()#while循环需要试试获取动态坐标变换，可能造成transform和time时间不统一
    ps.header.frame_id="turtle1"
    ps.point.x=2.0
    ps.point.y=3.0
    ps.point.z=5.0
    #转换逻辑实现，调用tf封装的算法
    # rospy.sleep(2)#通过休眠以接受完发布方的数据
    rate=rospy.Rate(10)#以1秒10次去接受发布方的数据
    while not rospy.is_shutdown():
        try:
            #转换实现
            ps_out=buffer.transform(ps,"world")#被转换的坐标点、目标坐标系
            #输出结果
            rospy.loginfo("转换后的坐标值:(%.2f,%.2f,%.2f),参考坐标系:%s",ps_out.point.x,ps_out.point.y,ps_out.point.z,ps_out.header.frame_id)
        except Exception as e:
            rospy.logwarn("错误信息:%s",e)
        
        rate.sleep()


    
    

File: transformation/src/dynamic/src/dynamic_s.cpp
#include "ros/ros.h"
#include "tf2_ros/transform_listener.h"//创建订阅对象所需包含的头文件
#include "tf2_ros/buffer.h"
#include "geometry_msgs/PointStamped.h"//组织坐标点数据所需包含的头文件
#include "tf2_geometry_msgs/tf2_geometry_msgs.h"//转换坐标系时所需包含的头文件

int main(int argc,char*argv[])
{
    setlocale(LC_ALL,"");//避免中文乱码问题
    ros::init(argc,argv,"dynamic_s");//初始化节点
    ros::NodeHandle nh;//创建节点句柄
    //创建订阅对象
    tf2_ros::Buffer buffer; //创建一个buffer缓存
    tf2_ros::TransformListener listenen(buffer);//再创建一个订阅对象【订阅对象可以将收到的数据放入缓存中】
    geometry_msgs::PointStamped ps;//组织一个坐标系数据【具体下方的消息设置参看坐标msgs消息】
    //参考坐标系
    ps.header.frame_id="turtle1";//小乌龟坐标系名称
    ps.header.stamp=ros::Time(0.0);
    ps.point.x=2.0;
    ps.point.y=3.0;
    ps.point.z=5.0;
    //转换算法，需要调用TF内置实现
    // ros::Duration(2).sleep();//添加休眠，以保证接收到发布方的消息
    ros::Rate rate(10);//1秒10次接受
    while(ros::ok())
    {   //核心代码——将ps转换成相对于base_link的坐标点
        geometry_msgs::PointStamped ps_out;
        try
        {
            ps_out=buffer.transform(ps,"world");//被转换坐标系点、转换的参考坐标系
            //输出
            ROS_INFO("转换后的坐标值:(%.2f,%.2f,%.2f),参考坐标系:%s",ps_out.point.x,ps_out.point.y,ps_out.point.z,ps_out.header.frame_id.c_str());
            
        }
        catch(const std::exception& e)
        {
            ROS_INFO("异常消息为:%s",e.what());
        }
        rate.sleep();
        ros::spinOnce();
    }
    
    return 0;
}